### DISCUSSION TITLE
fitHMM produces "Error in nlm"

### DISCUSSION THREAD

**Discussion by achance100** [bmcclintock/momentuHMM#66] [Q&A] [answered]
Hello,

I have successfully fit a 3 state HMM to relocation data from 35 individual deer with two covariates. Recently I received data from 6 additional individuals and have been unable to complete my re-analyses. The new data set is structurally the same as the old, however when I run the fitHMM function this error is returned:

#Error in nlm(nLogLike, optPar, nbStates, newformula, p$bounds, p$parSize,:non-finite value supplied by 'nlm'

In reading through the old moveHMM vingette it seems that this may be the result of covariate values that are too large. I standardized my covariate values and re-ran the function with no improvement. It seems unlikely to me that covariate values would be the problem in this situation because the same range of covariate values worked successfully on my original dataset. 

I have also tried playing around with my starting parameter values, but I've only been able to produce a new error:

#Error: Last global step failed to locate a point lower than x. Either x is an approximate local minimum of the function,
# the function is too non-linear for this algorithm, or steptol is too large.

Below is a summary of the code I'm currently using:

deerPrep <- prepData(tr, type = "UTM", coordNames = c("x","y"), covNames = c("dist.meters","tsh.hours"))

summary(deerPrep)

HMM data for 41 individuals:

100 -- 12736 observations
13 -- 12736 observations
17 -- 12736 observations
170 -- 11236 observations
20 -- 12736 observations
253 -- 12736 observations
255 -- 8932 observations
256 -- 10376 observations
259 -- 1590 observations
261 -- 12736 observations
262 -- 12736 observations
263 -- 12736 observations
265 -- 12736 observations
269 -- 12736 observations
27 -- 12736 observations
272 -- 5614 observations
273 -- 6724 observations
277 -- 12736 observations
278 -- 1621 observations
281 -- 12736 observations
284 -- 12736 observations
289 -- 12736 observations
293 -- 12736 observations
294 -- 12736 observations
295 -- 12736 observations
297 -- 12736 observations
299 -- 12737 observations
321 -- 12736 observations
329 -- 12736 observations
331 -- 11620 observations
337 -- 12736 observations
339 -- 12736 observations
341 -- 12736 observations
342 -- 12737 observations
344 -- 12736 observations
347 -- 12736 observations
3470 -- 12736 observations
348 -- 6724 observations
350 -- 12736 observations
97 -- 12736 observations
99 -- 12736 observations

Data summaries:

      step           angle       
 Min.   :   0    Min.   :-3.142  
 1st Qu.:   7    1st Qu.:-1.601  
 Median :  17    Median : 0.000  
 Mean   :  48    Mean   : 0.001  
 3rd Qu.:  54    3rd Qu.: 1.608  
 Max.   :2838    Max.   : 3.142  
 NA's   :10598   NA's   :14200 

# Initial parameters
mu0 <- c(7, 50, 150) # Mean step length
sigma0 <- c(10, 32, 100) # SD of the step length
zm0 <- c(0.001, 0, 0) # Basically split the difference in the zero-mass parameter for the proportion of step lengths that are expected to be zero in each state
am0 <- c(pi, 0, 0) # Turning angle mean
kappa0 <- c(0.1, 0.1, 0.1) # Turning angle concentration parameter (kappa between 0 and 1 for wrpcauchy)

#deerHMM3s <- fitHMM(deerPrep,
                    nbStates = 3, 
                    dist = list(step = "gamma",angle = "wrpcauchy"),
                    Par0 = list(step = c(mu0, sigma0,zm0), angle = c(am0, kappa0)),  
                    formula = ~tsh.hours+dist.meters, 
                    estAngleMean = list(angle=TRUE),
                    stationary = FALSE,
                    stateNames=stateNames,
                    nlmPar = list(print.level=2))

I'm looking for a way to troubleshoot these errors, specifically the first one mentioned:

#Error in nlm(nLogLike, optPar, nbStates, newformula, p$bounds, p$parSize,  : non-finite value supplied by 'nlm'

Thank you!

**Comment by bmcclintock** [comment, marked as answer]
Optimization issues such as this can be due to many things (e.g. poor initial parameter values, flat likelihood surfaces, parameters hitting boundaries, data errors, misspecified model) and can be tricky to diagnose. A "brute force" technique is to try many different starting values as described in this moveHMM [vignette](https://cran.r-project.org/web/packages/moveHMM/vignettes/moveHMM-starting-values.pdf) and/or different optimization routines (e.g. `optMethod="Nelder-Mead"`) and hope the problem goes away. It's also often helpful to start with simpler models, make sure they've converged, and then build up to more complex models (see #22).  

If the problem persists, then you can try and diagnose where the issue is occurring by examining the printed output returned by `fitHMM` using the argument `nlmPar=list(print.level=2)`.  When the `non-finite value supplied by 'nlm'` error occurs, there's often at least one parameter that is hitting a boundary (i.e. getting very large or very small). This can sometimes be mitigated using the `prior` argument to help prevent it from wondering into the nether regions of the parameter space. Another alternative is to fix the offending parameter to a very large or very small number (as appropriate) using the `fixPar` argument.

**Comment by achance100** [comment]
Hi Brett,

Thank you for pointing me towards the vignette on optimization routines. I have attempted to try many starting values using parallels as mentioned in part 6 of that vignette. However, the following error keeps occuring:

"Error in checkForRemoteErrors(val) :  7 nodes produced errors; first error: argument "Par0" is missing, with no default"

If I include a Par0 argument in fitHMM, then it prompts me to include a "dist" argument.

My code is as follows (for 3 behavioral states):

niter <- 25
			allPar0 <- lapply(as.list(1:niter), function(x) {
			stepMean0 <- runif(3,min = c(0.01,7,20),max = c(7,20,1200))
			stepSD0 <- runif(3,min = c(0.01, 7,20),max = c(7,20,1200))
			angleMean0 <- c(3,0,0)
			angleCon0 <- runif(3,min = c(0.1,1,2),max = c(1,2,5))
			stepPar0 <- c(stepMean0, stepSD0)
			anglePar0 <- c(angleMean0, angleCon0)
			return(list(step = stepPar0, angle = anglePar0))
			})

		allm_parallel <- parLapply(cl = cl, X = allPar0, fun = function(par0) {
				m <- fitHMM(data = deerPrep, 
								nbStates = 3, 
								stepPar0 = par0$step,
								anglePar0 = par0$angle,
								 dist = list(step = "gamma",angle = "wrpcauchy"))
				                                       return(m)})

The vignette example with 2 states doesn't show the Par0 argument being used, and indeed it seems redundant since the whole purpose of running multiple iterations is to find the appropriate starting values for step length and turning angles. Maybe I'm misinterpreting the purpose of the Par0 argument?

I will attempt to fix the parameter that is causing my initial error using the fixPar argument as you suggested, but trying multiple starting values appeals to me and I'd like to know how to make it work.

Thank you!

**Comment by bmcclintock** [comment]
I can't diagnose without a reproducible example, but it looks like you're confusing `moveHMM::fitHMM` and `momentuHMM::fitHMM`. `momentuHMM::fitHMM` always requires both `dist` and `Par0` to be specified (these arguments have no defaults), and there are no `stepPar0` or `anglePar0` arguments.

`momentuHMM::fitHMM` includes the `retryFits` argument for iteratively fitting the model using random perturbations of the current parameter estimates as the initial values (see `?momentuHMM::fitHMM`), but these are not done in parallel. To specify your own random (or grid search) initial values it does need to be done manually as you're trying to do above.
